### ISSUE TITLE
IS YOUR SERVER BROKEN SINCE 1.64 UPDATE? READ THIS

### ISSUE THREAD

**Issue by BoGuu** [AsYetUntitled/Framework#104]
Please make the changes from this commit: db66e545ef02faec2a2e9fd152eb5260b3deab8a, to your life_server/Functions/MySQL/fn_mresArray.sqf. If you use a headless client, change the relevant file in life_hc also.

You can thank @tkcjesse 

**Update**:

If you have malformed entries in your DB (contains ``) then replace your mresToArray with the following:

~~http://hastebin.com/omikofajej.cpp~~

https://gist.github.com/BoGuu/c552b652f752378f06a42c809150032e

Over time they should fix themselves, and you can revert back to the original.

**Comment by Jawshy** [comment]
More caps, please. :Kappa:

**Comment by setoy** [comment]
Hmm, have applied this to my server but the error still persists... Same error in the log files.

**Comment by tkcjesse** [comment]
I mean did you clear the column if you already joined?

**Comment by setoy** [comment]
Yeah, I've cleared the columns.

It appears on our modded server:
> 20:39:19 Error in expression <[[[''ACE_EarPlugs'',''ACE_packingBandage'',''>
20:39:19   Error position: <ACE_EarPlugs'',''ACE_packingBandage'',''>
20:39:19   Error Missing ]
20:39:19 Error in expression <[[[''ACE_EarPlugs'',''ACE_packingBandage'',''>
20:39:19   Error position: <ACE_EarPlugs'',''ACE_packingBandage'',''>
20:39:19   Error Missing ]
20:39:19 Error in expression <ay = call compile format["%1", _array];
_array;>
20:39:19   Error position: <_array;>
20:39:19   Error Undefined variable in expression: _array
20:39:19 File life_server\Functions\MySQL\fn_mresToArray.sqf, line 26

Aswell as this one on our clean one:

> 20:39:19 Error in expression <else{
_vehicle setFuel 1;
};

if (count _gear > 0 && (getNumber(missionConfigFil>
20:39:19   Error position: <_gear > 0 && (getNumber(missionConfigFil>
20:39:19   Error Undefined variable in expression: _gear
20:39:19 File life_server\Functions\Systems\fn_spawnVehicle.sqf, line 130
20:39:22 Error in expression <[''76561198013473651'',''76561198265154719'>
20:39:22   Error position: <76561198013473651'',''76561198265154719'>
20:39:22   Error Missing ]
20:39:22 Error in expression <[''76561198013473651'',''76561198265154719'>
20:39:22   Error position: <76561198013473651'',''76561198265154719'>
20:39:22   Error Missing ]
20:39:22 Error in expression <ay = call compile format["%1", _array];
_array;>
20:39:22   Error position: <_array;>
20:39:22   Error Undefined variable in expression: _array
20:39:22 File life_server\Functions\MySQL\fn_mresToArray.sqf, line 26
20:39:22 Error in expression < select 5] call DB_fnc_mresToArray;
if (_tmp isEqualType "") then {_tmp = call c>
20:39:22   Error position: <_tmp isEqualType "") then {_tmp = call c>
20:39:22   Error Undefined variable in expression: _tmp
20:39:22 File life_server\Functions\Gangs\fn_queryPlayerGang.sqf, line 16

**Comment by riskingfusion** [comment]
When ever I use that query code and execute it, it works but then when they re-join the server it re-appears with the '','' instead of ',' ?? Any fixes ?

**Comment by tkcjesse** [comment]
Tell us what columns are having the issue please.....

**Comment by tkcjesse** [comment]
If you're using the HC you need to update that file as well...

**Comment by tkcjesse** [comment]
@setoy  your error clearly says : DB_fnc_mresToArray

**Comment by vabene1111** [comment]
Any ideas what the reason for this could be: 

2016/09/22, 21:26:45 Error in expression <m 0 to (count _old)-1 do
{
_data = _old select _i;
_old set[_i,[_data select 0, >
2016/09/22, 21:26:45   Error position: <select _i;
_old set[_i,[_data select 0, >
2016/09/22, 21:26:45   Error Generic error in expression
2016/09/22, 21:26:45 File life_server\Functions\MySQL\fn_queryRequest.sqf, line 70`

We Updated the mresArray functions and restored the original Database (so no double quotes or anything broken) .. it kinda looks like the array has a different format ? running v4.4

**Comment by tkcjesse** [comment]
Dark has it working fine on 4.5 - https://i.gyazo.com/e3649b70802132e4c9faee2bb355bba1.png

**Comment by riskingfusion** [comment]
@tkcjesse I have never had a headless client file so I assume I am not using it. But is there any other ideas that may cause it to almost back track to the '','' instead of ',' even though I used that query code ?

**Comment by tkcjesse** [comment]
I mean what version are you using?... Look for the stupid str shit
Run it in debug to test if it's the issue and if it is, reform the data. Simple.

**Comment by riskingfusion** [comment]
We are using 4.4

**Comment by riskingfusion** [comment]
Dont have to be so rude if you cant be arsed to help fuck off!

**Comment by DarkSilencerZA** [comment]
It works fine for me. I'm using as close to the main build as possible, but just a ton of customized stuffs.

4.4 seems a little outdated.

**Comment by riskingfusion** [comment]
@DarkSilencerZA Yeah I had the issue when first updated I used the query code, It works then you disconnect re-log to the server and it is like it back tracks it self so you come back naked. So I am wondering if there was anything else you edited ?

**Comment by setoy** [comment]
> your error clearly says : DB_fnc_mresToArray

Yeah, but that file hasn't changed ;)

**Comment by tkcjesse** [comment]
Lmao. K.

**Comment by BoGuu** [comment]
If you are having problems post your mresArray and mresToArray, please

**Comment by setoy** [comment]
mresArray:
http://hastebin.com/fivakevexo.cpp

mresToArray:
http://hastebin.com/xemugacezi.cpp

**Comment by riskingfusion** [comment]
fn_mresArray.sqf
http://pastebin.com/A6zrQhXb

fn_mresToArray
http://pastebin.com/i0urCVYq

**Comment by SimZor** [comment]
I get these errors when joining after applying @tkcjesse fix.

http://hastebin.com/xivobukafi.bash

**Comment by ace16huey** [comment]
That SQL command is changing all the empty strings , `` to `, by the way.

**Comment by riskingfusion** [comment]
@ace16huey Some people are having the problem when they use that when you re-log to the server all them empty strings come back and make you naked again

**Comment by ace16huey** [comment]
@riskingfusion  I had a player that couldn't spawn in as a civ, I am gonna dump the table and replace them that way and hope for the best. I already messed up the licenses column by working to quick and not checking my work.

**Comment by riskingfusion** [comment]
@ace16huey Oh right. Thats why you make backups :P

**Comment by rebornfuel** [comment]
ant news o na fix or 4.4r3? as the current changes to fn_mresArray.sqf does not fix a thing

**Comment by ace16huey** [comment]
@riskingfusion Had a back up from a couple days ago, applied it and it got corrupted in the process... Arma updates give me the worst luck D:

**Comment by riskingfusion** [comment]
@rebornfuel not currently I think most of us are waiting on a fix for a similar issue

**Comment by riskingfusion** [comment]
@ace16huey  I think they give everyone bad luck xD

**Comment by harmdhast** [comment]
3.1.4.8 
players table fix.
```
UPDATE `players` 
 SET `cop_licenses` = replace(cop_licenses, '``', '`'),
 `civ_licenses` = replace(civ_licenses, '``', '`'),
 `med_licenses` = replace(med_licenses, '``', '`'),
 `aliases` = replace(aliases, '``', '`'),
 `cop_gear` = replace(cop_gear, '``', '`'),
 `civ_gear` = replace(civ_gear, '``', '`'),
 `med_gear` = replace(med_gear, '``', '`')
	WHERE `aliases` LIKE '"[``%'
```

**Comment by ace16huey** [comment]
That might break more then fix to be honest @harmdhast. did that to a couple columns and it cause all the empty strings to break and what not.

**Comment by harmdhast** [comment]
Please note the Where.

**Comment by riskingfusion** [comment]
@harmdhast I am going to test it on my server now

**Comment by rebornfuel** [comment]
dont bother it only fixes it for seconds, the issue comes back

**Comment by harmdhast** [comment]
The mresarray fix works like a charm.

**Comment by riskingfusion** [comment]
@harmdhast What version you on ?

**Comment by rebornfuel** [comment]
what version you on?

**Comment by rebornfuel** [comment]
4.4r3 it does nothing

**Comment by riskingfusion** [comment]
I am on 4.4 and it just comes back after re-log

**Comment by harmdhast** [comment]
@rebornfuel Are you sure you applied the https://github.com/AsYetUntitled/Framework/commit/db66e545ef02faec2a2e9fd152eb5260b3deab8a ?
Worked on 4.4r3 AND v3.1.4.8

**Comment by rebornfuel** [comment]
yeah its applied im taking my server down to test

**Comment by ace16huey** [comment]
@rebornfuel Okay, I see what you did now, the where would need to be in the selected column, 

`UPDATE `players` 
 SET `civ_gear` = replace(civ_gear, '``', '`')
WHERE `civ_gear` LIKE '"[``%'`

fixed it for that specific column. I don;t have much knowledge on SQL so thank you.

**Comment by harmdhast** [comment]
Basicly the SQL that I provided checks the aliases column if it is corrupted then fixes the entire line.

**Comment by harmdhast** [comment]
Since I only had a player table corrupted I can't provide any more scripts for other versions nor tables

**Comment by ace16huey** [comment]
Thats the only table I did do it to, I am about to give a link to 2 pics, then first being before I apply that code 

![](https://i.gyazo.com/1e66229d6481ade2fc1f548be47b7483.png)

then after


![](https://i.gyazo.com/b788d04a2c44831b8c19b0a2762bf6f6.png)


Can you see the issue? Its breaking the empty strings

**Comment by harmdhast** [comment]
Yeah, that's a v4 table right ? The script I provided is only for v3 since here you have to manage when it's [`` or Blaba`` 
And I don't have a table to test so, I only did it for v3. Or you "fixed" it uncompletely and my script fucks up.

**Comment by ace16huey** [comment]
This is a v3 table. 3.1.4.8 to be exact. But I will look into it more

**Comment by harmdhast** [comment]
That's strange did you tried to fix it before applying my script ? It's not supposed to be only double quotes.

**Comment by ace16huey** [comment]
I have done so many things to it, that I lost track, I am gonna revert and apply the fix again to see if I messed something up

**Comment by harmdhast** [comment]
And sadly Mysql does not support regex replace.
That why you don't make databases like a web developper when you make a mission used by a whole fukin community.

**Comment by setoy** [comment]
You could dump the DB and do regex replaces in Notepad++ or such (:

**Comment by harmdhast** [comment]
Yeah but this shouldn't happen. And ppl should be aware of encoding and such things with dumping dbs. Just to say that you never dumb a database to replace string. Just think about 10k entries databases (and that's not much)

**Comment by angryfatty** [comment]
we got the database fixed but we have a problem with whitelisting as cops the host can join as cop but when I join on the server after whitelisting I spawn naked without spawn screen but when I go back to civ it works fine any suggestions

**Comment by BoGuu** [comment]
@setoy what exactly is your error?
@riskingfusion You haven't applied the fix

**Comment by harmdhast** [comment]
@angryfatty
You didn't fixed the cop column. Check the integrity of the cop column or reset it to the default 
```
"[]"
```

**Comment by JedINyte** [comment]
Running a 4.4 server (not r3) none of the fixes above fix the issue of spawning with no gear or licenses :( It initially works, but as soon as you leave the server and rejoin (lobby or server list) the columns re write the '' into the database.

**Comment by BoGuu** [comment]
@JedINyte Post your mres (array) functions, and a snap shot of your DB.

**Comment by JedINyte** [comment]
Hello Boguu, im fairly new to all this, never had to delve this deep into the database before: mresarray:


` /*
     File: fn_mresArray.sqf
     Author: Bryan "Tonic" Boardwine";
 
     Description:
     Acts as a mres (MySQL Real Escape) for arrays so they
     can be properly inserted into the database without causing
     any problems. The return method is 'hacky' but it's effective.
  */
  private ["_array"];
  _array = [_this,0,[],[[]]] call BIS_fnc_param;
  _array = str _array;
  _array = toArray(_array);
  
  for "_i" from 0 to (count _array)-1 do
 {
     _sel = _array select _i;
     if ((_i != 0 && _i != ((count _array)-1))) then
     {
         if (_sel isEqualTo 34) then
         {
             _array set[_i,96];
         };
      };
  };

  str(toString(_array));`

mres to array:

`/*
    File: fn_mresToArray.sqf
    Author: Bryan "Tonic" Boardwine";
    Description:
    Acts as a mres (MySQL Real Escape) for arrays so they
    can be properly inserted into the database without causing
    any problems. The return method is 'hacky' but it's effective.
*/
private["_array"];
_array = [_this,0,"",[""]] call BIS_fnc_param;
if (_array isEqualTo "") exitWith {[]};
_array = toArray(_array);

for "_i" from 0 to (count _array)-1 do
{
    _sel = _array select _i;
    if (_sel == 96) then
    {
        _array set[_i,39];
    };
};

_array = toString(_array);
_array = call compile format["%1", _array];
_array;`

what do you mean by a snapshot of DB?

**Comment by tkcjesse** [comment]
I assume the ' at the start of that file and end on jedi is an attempt at markdown...  

Also you guys need to make sure you're not doing the sql update while that player is on the server.... hope thats obvious...

**Comment by tkcjesse** [comment]
Hope we can all now learn why its important to test arma updates on a test instance and db prior to updating a live production server... and why doing a backup is necessary before hand.

**Comment by JedINyte** [comment]
The ' at the start and end is where i tried to add it as code on here lol (the little <> box)!! I thought it would put it in a neat box, but it just displayed as is with the' at each end. The player isn't on the server when updating the SQL :)

**Comment by jmuirhead** [comment]
After the update i get this error now. Running 4.4r3
`20:13:39 "------------------------------------------------------------------------------------------------------"
20:13:39 "--------------------------------- Starting Altis Life Client Init ----------------------------------"
20:13:39 "------------------------------------------------------------------------------------------------------"
20:13:39 "::Life Client:: Initialization Variables"
20:13:39 "::Life Client:: Variables initialized"
20:13:39 "::Life Client:: Setting up Eventhandlers"
20:13:39 "::Life Client:: Eventhandlers completed"
20:13:39 "::Life Client:: Setting up user actions"
20:13:39 "::Life Client:: User actions completed"
20:13:39 "::Life Client:: Waiting for server functions to transfer.."
20:13:39 "::Life Client:: Received server functions."
20:13:39 "::Life Client:: Waiting for the server to be ready.."
20:13:40 Error in expression <e["_itemArray","_handle"];
_itemArray = life_gear;
waitUntil {!(isNull (findDisp>
20:13:40   Error position: <life_gear;
waitUntil {!(isNull (findDisp>
20:13:40   Error Undefined variable in expression: life_gear
20:13:40 File mpmissions\__CUR_MP.Altis\core\functions\fn_loadGear.sqf, line 10
20:13:41  Mission id: 917ba15411c127d9e9fe6ec6551f773335cc72b0
20:13:41 Wrong color format size #FFFFF
20:13:41 Error in expression <itUntil {scriptDone _handle};

if(count _itemArray isEqualTo 0) exitWith {
switc>
20:13:41   Error position: <_itemArray isEqualTo 0) exitWith {
switc>
20:13:41   Error Undefined variable in expression: _itemarray
20:13:41 File mpmissions\__CUR_MP.Altis\core\functions\fn_loadGear.sqf, line 16
20:13:41 Warning: no type entry inside class life_spawn_selection/controls/LineMarker 
20:13:44 "Past Settings Init"
20:13:44 "Executing client.fsm"
20:13:44 "Display 46 Found"
20:13:44 "------------------------------------------------------------------------------------------------------"
20:13:44 "                End of Altis Life Client Init :: Total Execution Time 4.66211 seconds "
20:13:44 "------------------------------------------------------------------------------------------------------"
`

**Comment by setoy** [comment]
@boguu 4.4R3

**Comment by Jawshy** [comment]
### If one is going to post code then please remember to format it. See [Mastering Markdown](https://guides.github.com/features/mastering-markdown/) → Examples → Code. 

> \`\`\`
> if (isAwesome){
>   return true
> }
> \`\`\`

### If one is going to paste _a lot_ of code then please upload it as a file to GitHub instead.

**Comment by Jawshy** [comment]
@JedINyte, one backtick <code>\`</code> is for single-line code. Three backticks <code>\`\`\`</code> above and below are more multi-line code blocks. Hope that clears it up.

**Comment by Gentism** [comment]
im still having the issue with db after doing the mresarray fix anyone help??

**Comment by harmdhast** [comment]
@Gentism You need to fix your database as well. Either restore a backup or look for some little scripts in this issue thread.

**Comment by SimZor** [comment]
@rebornfuel as @tkcjesse said. Why it only works for seconds is because you still have players on your server while updating the MySQL entries.

Just shut your server down, update the tables and then connect and it will work. At least for that issue.

**Comment by xProlithium** [comment]
But we still haven't pinpointed the issue itself or what is causing it?

**Comment by Jawshy** [comment]
@xProlithium, [str](https://community.bistudio.com/wiki/str) on the Bohemia Interactive Community Wiki: 

> ... Since Arma 3 v1.63.137891 this command will correctly escape all " inside a string: 

[str command does not escape nested double-quotes for strings](https://feedback.bistudio.com/T79720) on the Bohemia Interactive Feedback Tracker.

**Comment by SimZor** [comment]
I just stumbled upon another issue and that is that as soon as you open your house crates everything and then log back in everything is gone.

See error log http://hastebin.com/xivobukafi.bash

**Comment by Jawshy** [comment]
@SimZor, it is the same issue; it is just affecting a different table. 

----

Reference: https://github.com/AsYetUntitled/Framework/issues/81

**Comment by xProlithium** [comment]
I have an older version of tonic's mres so not only am I somewhat lost and how to fix it, I feel in the dark due to the version of the files as well.

**Comment by Jawshy** [comment]
@xProlithium, attach a copy of your [fn_mresArray.sqf](https://github.com/AsYetUntitled/Framework/blob/master/life_server/Functions/MySQL/fn_mresArray.sqf) to a comment.

**Comment by xProlithium** [comment]
@Jawshy
`private["_array"];
_array = [_this,0,[],[[]]] call BIS_fnc_param;
_array = str(str(_array));
_array = toArray(_array);

for "_i" from 0 to (count _array)-1 do
{
_sel = _array select _i;
if((_i != 0 && _i != ((count _array)-1))) then
{
if(_sel == 34) then
{
_array set[_i,96];
};
};
};

toString(_array);`

**Comment by Jawshy** [comment]
https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249102698

**Comment by xProlithium** [comment]
I tried Insert Code, know what it didn't do? Insert code <_<

**Comment by xProlithium** [comment]
Better? I tried it the first time with ' and it didn't work lol

**Comment by SimZor** [comment]
@Jawshy so that means I can just do the DB fix for that table aswell?

**Comment by setoy** [comment]
I needed to fix the following columns in my DB:

containers:
inventory
gear

gangs:
members

players:
aliases
civ_licenses
cop_licenses
med_licenses
civ_gear
cop_gear
med_gear

vehicles:
inventory
gear

wanted
wantedCrimes

**Comment by Jawshy** [comment]
@xProlithium, 

```sqf
private["_array"];
_array = [_this,0,[],[[]]] call BIS_fnc_param;
_array = str _array;
_array = toArray(_array);

for "_i" from 0 to (count _array)-1 do
{
_sel = _array select _i;
if((_i != 0 && _i != ((count _array)-1))) then
{
if(_sel == 34) then
{
_array set[_i,96];
};
};
};

str(toString(_array));
```

**Comment by Jawshy** [comment]
@SimZor, yes. See @setoy's [comment above](https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249122182).

**Comment by SimZor** [comment]
There was no ´´ in my DB in any other tables than players, but yet it sometimes spits out the error.

@Jawshy

**Comment by xProlithium** [comment]
@Jawshy I'm getting an error in the asyncCall.sqf on line 53 and 56.
```
private["_queryStmt","_queryResult","_key","_mode","_return","_loop"];

_queryStmt = [_this,0,"",[""]] call BIS_fnc_param;
_mode = [_this,1,1,[0]] call BIS_fnc_param;
_multiarr = [_this,2,false,[false]] call BIS_fnc_param;

_key = "extDB2" callExtension format["%1:%2:%3",_mode,(call life_sql_id), _queryStmt];

if(_mode isEqualTo 1) exitWith {true};

_key = call compile format["%1",_key];
_key = _key select 1;

uiSleep (random .03);

_queryResult = "";
_loop = true;
while{_loop} do
{
	_queryResult = "extDB2" callExtension format["4:%1", _key];
	if (_queryResult isEqualTo "[5]") then {
		_queryResult = "";
		while{true} do {
			_pipe = "extDB2" callExtension format["5:%1", _key];
			if(_pipe isEqualTo "") exitWith {_loop = false};
			_queryResult = _queryResult + _pipe;
		};
	}
	else
	{
		if (_queryResult isEqualTo "[3]") then
		{
			uisleep 0.1;
		} else {
			_loop = false;
		};
	};
};
if(typeName _queryResult == "STRING") then {_queryResult = call compile _queryResult};

_return = (_queryResult select 1);

if(!_multiarr) then {
	_return = _return select 0;
} else {
	if((count (_return select 0)) == 0) then {_return = []};
};

_return;
```

Not sure what it refers to.

**Comment by BartoCarto** [comment]
I'm using 3.1.4.8
i have problem when players log out there gear is disappear after this update
how i can solve this?

**Comment by tkcjesse** [comment]
Lmfaoooo. Ong

**Comment by SimZor** [comment]
Is this troll?

Didn't they teach you to read in school?

**Comment by tkcjesse** [comment]
Dude HAS to be trolling. Holy fuck. Man this thread... lmao.

**Comment by BoGuu** [comment]
So is any one still broken?

**Comment by Twisted123** [comment]
yes we are broken still

**Comment by tkcjesse** [comment]
So @BoGuu whats been the issues? Does this actually not fix it for some peoplem or is it bad db stuff after they've realized somethings broken?

**Comment by BartoCarto** [comment]
I've tried all the written

mresArray
http://pastebin.com/YD0UsEm5

mresToArray
http://pastebin.com/9ZAR5XT7

**Comment by Jawshy** [comment]
https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249028317 

https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249198252

@riskingfusion and @BartoCarto, your [fn_mresArray.sqf](https://github.com/AsYetUntitled/Framework/blob/master/life_server/Functions/MySQL/fn_mresArray.sqf) file **has not been updated** with [this change](https://github.com/AsYetUntitled/Framework/commit/db66e545ef02faec2a2e9fd152eb5260b3deab8a).

**Comment by BoGuu** [comment]
@Twisted123 Please post the files asked for, and a screenshot of your database (with recently logged in people)

@BartoCarto You haven't followed the one thing you were told to do in the OP

**Comment by BartoCarto** [comment]
i have made this change https://github.com/AsYetUntitled/Framework/commit/db66e545ef02faec2a2e9fd152eb5260b3deab8a

same problem

**Comment by Jawshy** [comment]
@BartoCarto, the fn_mresArray.sqf on the Pastebin file you linked does not have the latest change. The green highlight on the diff is the latest code.

**Comment by Jawshy** [comment]
If your database contains malformed entries then following the steps outlined in https://github.com/AsYetUntitled/Framework/issues/104#issue-178678747. 

> If you have malformed entries in your DB, run the following command:

> ```sql
> UPDATE `players`
> SET `civ_licenses` = replace(civ_licenses, '``', '`') where uid>=0;
> ```

> Replace `players` and `civ_licenses` with relevant table/column (licenses, gear, containers etc etc.)

> Or alternatively you can dump your DB, and find/replace \`\` -> `, then reupload.

**Comment by Twisted123** [comment]
Has this been fixed yet or anyone found a way to fix it?

**Comment by BoGuu** [comment]
@Twisted123 what exactly is your problem? You haven't posted any details about what your particular problem is, have you even made the changes stated?

**Comment by numpty9989** [comment]
after running this "fix" got https://gyazo.com/478e2e9eb6cc58588a386e838e716690 any help would be appreciated

**Comment by Emma-L** [comment]
@numpty9989 Post a picture of your gear in the database

**Comment by numpty9989** [comment]
https://gyazo.com/b13b6fad99b0f37dcb0e6d6e1a6ba3f0

**Comment by Emma-L** [comment]
@numpty9989 Please refer to the original post, which contains SQL to remove the malformed entries. (The double \`\`)

**Comment by numpty9989** [comment]
me running query https://gyazo.com/709163b4ba3cd810f959a54e165181f2
me running it https://gyazo.com/2995b13a11d01eec3c0d6b5bbdb78e20
original time i did it said affected 24 rows

**Comment by Emma-L** [comment]
@numpty9989 'Replace players and civ_licenses with relevant table/column (licenses, gear, containers etc etc.)'

**Comment by numpty9989** [comment]
huh

**Comment by numpty9989** [comment]
@setoy you linked to the same issue xD

**Comment by LiamLikesGaming** [comment]
I have done the fn_mresArray.sqf fix, but now when i try to join a job I get stuck on "Sending request to server for player information UID". Do I need to do the table fix too? Or what?

**Comment by numpty9989** [comment]
@Emma-L  Fixed Now thanks i had a noob moment

**Comment by Emma-L** [comment]
@LiamLikesGaming Like any issue, you need to post more details/logs. Check if your DB has any double back-ticks, like this: https://i.gyazo.com/35570faa0aa355f2cbe0d0ff85d3c9d4.png. If it does, refer back to the first post.

@numpty9989 No worries, glad you got it fixed!

**Comment by LiamLikesGaming** [comment]
I don't really know I have, I'm fairly new to MYSQL, and don't really understand much of it.

Here is a screenshot anyway. I can't tell if it has got any "double back-ticks" or not.

**Comment by Emma-L** [comment]
> Here is a screenshot anyway

You forgot to attach it :)

**Comment by numpty9989** [comment]
#epic fail lol

**Comment by LiamLikesGaming** [comment]
I edited the comment.

**Comment by Emma-L** [comment]
@LiamLikesGaming Run this:

```sql
UPDATE `players` 
 SET `cop_licenses` = replace(cop_licenses, '``', '`'),
 `civ_licenses` = replace(civ_licenses, '``', '`'),
 `med_licenses` = replace(med_licenses, '``', '`'),
 `aliases` = replace(aliases, '``', '`'),
 `cop_gear` = replace(cop_gear, '``', '`'),
 `civ_gear` = replace(civ_gear, '``', '`'),
 `med_gear` = replace(med_gear, '``', '`')
    WHERE uid>=0;```

**Comment by Twisted123** [comment]
where do we run it?

**Comment by Emma-L** [comment]
....In MySQL workbench (or whatever you use)...to set up your server you must have had to run some SQL code...

**Comment by LiamLikesGaming** [comment]
Error Code: 1064. You have an error in your SQL syntax; check the manual that corresponds to your MYSQL server version for the right syntax to use near '``' at line 1.

Thats the error it gives when I used that code, however when I ran it, it saids "Changed: 4" so it must've done something

**Comment by Emma-L** [comment]
@LiamLikesGaming I fucked my markdown up, sorry my triple back-ticks at the end were by accident. Did you include them. If you did rerun it (I edited the comment to get rid of them).

**Comment by LiamLikesGaming** [comment]
I did that command too, even the edited one and I still get "Sending request to server for player information UID", sorry for this by the way..

**Comment by Emma-L** [comment]
Can you verify all of your DB for back-ticks (\`\`) and post logs? Alternatively, if you don't have too much data in your DB, and you're not too bothered about it, you can simply reset it.

**Comment by LiamLikesGaming** [comment]
How do I verify the DB for back-ticks. And I have quite a lot of data for players if your talking about resetting it..

Thanks

**Comment by Emma-L** [comment]
You literally just look for double back-ticks (/`/`)

**Comment by LiamLikesGaming** [comment]
might it be easier if I can contact you somewhere and give you access to the DB itself?

Thank you.

**Comment by Emma-L** [comment]
You can contact me on the, either, the AYU discord or the ArmaRPGLife discord (see the readme for an inv). If I'm not around, as I'm going off soon, there will be other people there that can help you, who are probably more proficient anyway.

**Comment by LiamLikesGaming** [comment]
Okay i added you on discord.

**Comment by variousskittles** [comment]
So I edit the fn_mresArray.sqf with the db66e54 ?

**Comment by xProlithium** [comment]
I did some troubleshooting. The error I get after changing mresArray for version 3.1.4.8 doesn't happen in the version 4.4r3 after changing the mresArray.

**Comment by variousskittles** [comment]
Im getting this message before and after doing this, I dont know if the message is the same as everyone else or not but it says  "The server didnt find any any player information matching your UID, attempting to add player to system." when joining

**Comment by Jawshy** [comment]
https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249293382 

@numpty9989, no he did not. Someone posted a duplicate issue (of this one) and @setoy [redirected them](https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests) to this issue. GitHub then referenced back to the duplicate issue. 

----

@variousskittles, your character's data is being sent to the server and added to the database. This should only occur once on the initial setup.

**Comment by numpty9989** [comment]
@Jawshy  my bad soz

**Comment by harmdhast** [comment]
Here you go, the complete SQL fix for the **v3.1.4.8**.

### YOUR SERVER MUST BE STOPPED OR YOU MAY CORRUPT THE WHOLE FIX. IF ITS THE CASE PLEASE DO A BACKUP OR FIX YOU DATABASE MANUALLY

```
UPDATE `players`
 SET `cop_licenses` = replace(cop_licenses, '``', '`'),
 `civ_licenses` = replace(civ_licenses, '``', '`'),
 `med_licenses` = replace(med_licenses, '``', '`'),
 `aliases` = replace(aliases, '``', '`'),
 `cop_gear` = replace(cop_gear, '``', '`'),
 `civ_gear` = replace(civ_gear, '``', '`'),
 `med_gear` = replace(med_gear, '``', '`')
    WHERE `aliases` LIKE '"[``%';
   
UPDATE `gangs`
 SET `members` = replace(members, '``', '`')
    WHERE `members` LIKE '"[``%';
   
UPDATE `houses`
 SET `containers` = replace(containers, '``', '`')
    WHERE `containers` LIKE '"[[``%';
   
UPDATE `houses`
 SET `inventory` = replace(inventory, '``', '`')
    WHERE `inventory` LIKE '"[[[``%';
```

**Comment by variousskittles** [comment]
@Jawshy It happens everytime I join my server.

**Comment by harmdhast** [comment]
SQL fix for the v4+. **STILL WIP**

YOUR SERVER MUST BE STOPPED OR YOU MAY CORRUPT THE WHOLE FIX. IF ITS THE CASE PLEASE DO A BACKUP OR FIX YOU DATABASE MANUALLY

```
UPDATE `players`
 SET `cop_licenses` = replace(cop_licenses, '``', '`'),
 `aliases` = replace(aliases, '``', '`'),
 `cop_gear` = replace(cop_gear, '``', '`')
    WHERE `cop_licenses` LIKE '"[[``%';
    
UPDATE `players`
 SET `civ_licenses` = replace(civ_licenses, '``', '`'),
 `civ_gear` = replace(civ_gear, '``', '`'),
 `aliases` = replace(aliases, '``', '`')
    WHERE `civ_licenses` LIKE '"[[``%';
    
UPDATE `players`
 SET `med_licenses` = replace(med_licenses, '``', '`'),
 `aliases` = replace(aliases, '``', '`'),
 `med_gear` = replace(med_gear, '``', '`')
    WHERE `med_licenses` LIKE '"[[``%';
   
UPDATE `gangs`
 SET `members` = replace(members, '``', '`')
    WHERE `members` LIKE '"[``%';
   
UPDATE `containers`
 SET `gear` = replace(gear, '``', '`')
    WHERE `gear` LIKE '"[[``%';
   
UPDATE `containers`
 SET `inventory` = replace(inventory, '``', '`')
    WHERE `inventory` LIKE '"[[[``%';
    
UPDATE `vehicles`
 SET `gear` = replace(gear, '``', '`')
    WHERE `gear` LIKE '"[[``%';
   
UPDATE `vehicles`
 SET `inventory` = replace(inventory, '``', '`')
    WHERE `inventory` LIKE '"[[[``%';
```

**Comment by Jawshy** [comment]
@variousskittles, your issue appears to be unrelated to the one listed here. If you like you can join the [ArmA RPG Life Discord server](https://discordapp.com/invite/sEcQdPn) or post on the [forums](https://armarpglife.com/forums/) where others can help you troubleshoot your issue.

**Comment by BoGuu** [comment]
See updated OP:

`If you have malformed entries in your DB (contain ``) then replace your mresToArray with the following:

http://hastebin.com/onemijeyab.cpp

Over time they should fix themselves, and you can revert back to the original.`

**Comment by harmdhast** [comment]
Useless. Better fix it right now. Temp fixes doesn't go well with AltisLife and Arma.
One thing to note, why does the database is still managed like that ?
It look like one of the web database that you do in the html tutorials. Like seriously. One does not store arrays as strings. And it's just one of the things.

**Comment by BoGuu** [comment]
> Useless. Better fix it right now. Temp fixes doesn't go well with AltisLife and Arma.
Unfortunately, like with anything in AltisLife, a large amount of people fail to follow simple instructions. This has been a perfect demonstration of this, **countless** people in the same boat. And in the same fashion as usual, it's simplified for people.

> One thing to note, why does the database is still managed like that ?

Because it's legacy.

> It look like one of the web database that you do in the html tutorials. Like seriously. One does not store arrays as strings. And it's just one of the things.

The system has not been changed in a long time, but not too long ago there was a reasoning behind it, but by saying that, you come across as naïve.

**Comment by CarbineLife** [comment]
Ok so I bought a server right before the 1.64 update hit so we had nothing on it no player data everything was just standard brand new altis life version 4.4. The day after however the 1.64 update hit so we updated to it and also updated our altis life version to 4.5. When we did that every time we try and join the server now we get these messages constantly alternating between eachother on a black screen.

The server didn't find any player information matching your UID, attempting to add player to system.

Sending request to server for player information UID [76561198096865516]

Can anyone link me to a fix for this or just tell me on here. We've tried contacting the hosting company and they said wait for an update on the Altis Life mod however I don't see signs of an upate coming so what is going on? Please someone just help.

**Comment by Jawshy** [comment]
@CarbineLife, your issue does not appear to be related to this one. Please jump in #server_setup on the [ArmA RPG Life Discord server](https://discordapp.com/invite/sEcQdPn) for support.

**Comment by CarbineLife** [comment]
@Jawshy will do thanks for the info

**Comment by 3LGStevo** [comment]
Neither the MresArray.sqf file or MresToArray.sqf file changes have fixed the issue. When my players still sync their data, it still inserts `` into the database.

**Comment by BoGuu** [comment]
@3LGStevo 

It would seem you have not correctly edited the files then. The only cause for the double \`\` would be by not changing the files correctly. You should have some \`\` though, in between commas, such as:

```csharp
[`XXXX`,``,``,`XXXX`]
```

**Comment by chelseaboyc18** [comment]
@BoGuu  im having same problem as 3LGStevo it allows me into server and can relog as many times as i want but soon as i sync data and then relog i get received request from server validating

**Comment by chelseaboyc18** [comment]
Or even items or licences anythink to do with arrays the way arrays are inserted into DB is not correct, they have to rewrite the two functions that format arrays OR remove em completely because they are not needed

**Comment by 3LGStevo** [comment]
Yeah, someone posted a "fix" in one of the replies which didn't fix it, and the hastebin links weren't working for me.

@chelseaboyc18
I'll post you the files tomorrow, you'll need to repair your database entries. Better to take your server down to prevent further corruptions.

**Comment by Emma-L** [comment]
@3LGStevo The gist link is working just fine for me, the hastebin link has been crossed out for a reason..

**Comment by BoGuu** [comment]
@chelseaboyc18 

> Or even items or licences anythink to do with arrays the way arrays are inserted into DB is not correct

It will be correct after applying the mresArray changes in the OP (correct as in pre update).

> they have to rewrite the two functions that format arrays OR remove em completely because they are not needed

One has been, as you know, the other doesn't need to change seeing as the format will be no different. However, you are right in saying they are not needed anymore (and haven't really been for a long time).

> Yeah, someone posted a "fix" in one of the replies which didn't fix it, and the hastebin links weren't working for me.

Gist link works, although it's not a fix, it's a hack to fix database entries.

@chelseaboyc18 @3LGStevo 

The changes have worked for servers running every version there is, so if you still have persistent problems then you've either still got malformed arrays (and no mresToArray changes), the malformed arrays have caused null entries, or your problem is something specific to your setup.

**Comment by 3LGStevo** [comment]
@BoGuu 
My server has been working for a few days now, dude. I thought the fix didn't work because someone posted a "fix" using pastebin which did nothing to rectify the issue. The hastebin links didn't work (and are now crossed out) but the working link was added in afterwards.

Anyway, @chelseaboyc18 , replace the files in life_server\functions\MySQL\ with the two files below. Any database entries with `` will need to be manually corrected. I say manually, because each of the automated methods above just fucks a different part of the database for that record, so just have some patience and work through them manually. It's easier if you filter the table with the records that are broken (like searching for '%[``%').

MREStoArray.sqf
`/*
	File: fn_mresToArray.sqf
	Author: Bryan "Tonic" Boardwine";
	
	Description:
	Acts as a mres (MySQL Real Escape) for arrays so they
	can be properly inserted into the database without causing
	any problems. The return method is 'hacky' but it's effective.
*/
private["_array"];
_array = [_this,0,"",[""]] call BIS_fnc_param;
if (_array isEqualTo "") exitWith {[]};
_array = toArray(_array);
 
for "_i" from 0 to (count _array)-1 do
{
    _sel = _array select _i;
    if (_sel == 96) then
    {
        _array set[_i,39];
    };
};
 
_array = toString(_array);
_array = call compile format["%1", _array];
_array;
`

MRESArray.sqf
`/*
	File: fn_mresArray.sqf
	Author: Bryan "Tonic" Boardwine";
	
	Description:
	Acts as a mres (MySQL Real Escape) for arrays so they
	can be properly inserted into the database without causing
	any problems. The return method is 'hacky' but it's effective.
*/
private["_array"];
_array = [_this,0,[],[[]]] call BIS_fnc_param;
_array = str(_array);
_array = toArray(_array);
 
for "_i" from 0 to (count _array)-1 do
{
    _sel = _array select _i;
    if ((_i != 0 && _i != ((count _array)-1))) then
    {
        if (_sel isEqualTo 34) then
        {
            _array set[_i,96];
        };
    };
};
 
str(toString(_array));`

**Comment by Jawshy** [comment]
@3LGStevo, code blocks are formatted with three backticks <code>`</code> above and below; not one. However, with the amount of code you have there you should upload the files to GitHub instead as said at https://github.com/AsYetUntitled/Framework/issues/104#issuecomment-249102698.

**Comment by chelseaboyc18** [comment]
@BoGuu  thanks for that finally got it working now. Appreciate your help.
Also i noticed in your MREStoArray.sqf  you have  str(toString(_array));` should be str(toString(_array));
just pointing out as anyone else might make the mistake of adding the extra ' 
(:

**Comment by harmdhast** [comment]
@chelseaboyc18 
It seems to be markdown related.

**Comment by VladmirVorkeshky** [comment]
Can someone please help me with this? I am new to GitHub and would prefer if someone could help me through Steam @Vladmir Vorkeshky. I did everything he said to do and I still am having the same problem. I have a community and they're all waiting on me to fix it I would REALLY REALLY appreciate it if anyone could help. Thanks!

**Comment by OfficialTemp** [comment]
@VladmirVorkeshky Make sure you have followed ALL the steps listed by Boguu. If you still cannot fix it, join the Altis Life RPG discord and ask for help by someone there. https://discordapp.com/invite/sEcQdPn

**Comment by OfficialTemp** [comment]
@VladmirVorkeshky I can also help you fix it if you want, if no-one else can.

**Comment by JimBob0121** [comment]
UPDATE `players`
SET `cop_licenses` = replace(cop_licenses, '``', '`'),
`aliases` = replace(aliases, '``', '`'),
`cop_gear` = replace(cop_gear, '``', '`')
WHERE `cop_licenses` LIKE '"[[``%';

UPDATE `players`
SET `civ_licenses` = replace(civ_licenses, '``', '`'),
`civ_gear` = replace(civ_gear, '``', '`'),
`aliases` = replace(aliases, '``', '`')
WHERE `civ_licenses` LIKE '"[[``%';

UPDATE `players`
SET `med_licenses` = replace(med_licenses, '``', '`'),
`aliases` = replace(aliases, '``', '`'),
`med_gear` = replace(med_gear, '``', '`')
WHERE `med_licenses` LIKE '"[[``%';

UPDATE `gangs`
SET `members` = replace(members, '``', '`')
WHERE `members` LIKE '"[``%';

UPDATE `containers`
SET `gear` = replace(gear, '``', '`')
WHERE `gear` LIKE '"[[``%';

UPDATE `containers`
SET `inventory` = replace(inventory, '``', '`')
WHERE `inventory` LIKE '"[[[``%';

UPDATE `vehicles`
SET `gear` = replace(gear, '``', '`')
WHERE `gear` LIKE '"[[``%';

UPDATE `vehicles`
SET `inventory` = replace(inventory, '``', '`')
WHERE `inventory` LIKE '"[[[``%';

**Comment by JamesMason580** [comment]
Really don't want to sound like a broken record, but I am comparatively new to running an Arma 3 server. We're using the old CG mission file with a few modifications of our own. Now I've run into this issue, gear saving works perfectly, but when anyone loads in after having saved gear, they spawn in naked. A temporary fix I instituted was to if(count _itemArray == 0) exitWith within the loadGear.sqf to if(count 0 == 0) exitWith so that it always evaluates to true and give everyone the default gear. My problem is that I'm unsure how to debug this, or I could probably figure it out. Any help would be greatly appreciated.

**Comment by wilsonjude** [comment]
So there isnt a fix?

**Comment by NeFShARk** [comment]
@wilsonjude Did you read this issue at least?

**Comment by wilsonjude** [comment]
Now it says Setting Up Client, Please wait

**Comment by Emma-L** [comment]
@wilsonjude if you have followed all of the steps and still can't get it working, please join the ArmA RPG Life discord: [invite.](https://discord.gg/g4DkYY2)
